File: app/src/main/java/com/stan/dagger_android/news/NewsPresenter.java
package com.stan.dagger_android.news;

import com.stan.dagger_android.base.DisposablePresenter;
import com.stan.dagger_android.data.BaseBackData;
import com.stan.dagger_android.net.Api;
import com.stan.dagger_android.util.RxUtils;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * Created by uu on 2017/12/29.
 */

public class NewsPresenter extends DisposablePresenter implements ActivityNewsContract.Presenter {

    public static final String TAG = "NewsPresenter";

    //view not inject
    ActivityNewsContract.View view;

    @Inject
    Api retroApi;

    @Inject
    public NewsPresenter() {
    }


    @Override
    public void takeView(ActivityNewsContract.View view) {
        this.view = view;
    }

    @Override
    public void dropView() {
        clearDisposable();
        this.view = null;
    }

    @Override
    public void getMessage(String message) {
        retroApi.home()
                .compose(RxUtils.<BaseBackData>applySchedulers())
                .subscribe(new Observer<BaseBackData>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        addDisposable(d);
                    }

                    @Override
                    public void onNext(BaseBackData baseBackData) {
                        view.showMessage(baseBackData.message);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/stan/dagger_android/news/NewsPresenter.kt
package com.stan.dagger_android.news

import com.orhanobut.logger.Logger
import com.stan.dagger_android.base.DisposablePresenter
import com.stan.dagger_android.data.BaseBackData
import com.stan.dagger_android.net.Api
import com.stan.dagger_android.util.RxUtils
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import javax.inject.Inject

/**
 * Created by uu on 2017/12/29.
 */

class NewsPresenter @Inject
constructor() : DisposablePresenter(), ActivityNewsContract.Presenter {

    internal var view: ActivityNewsContract.View? = null

    @Inject
    lateinit var retroApi: Api


    override fun takeView(view: ActivityNewsContract.View) {
        this.view = view
    }

    override fun dropView() {
        clearDisposable()
        this.view = null
    }

    override fun getMessage(message: String) {
        retroApi.home()
                .compose(RxUtils.applySchedulers())
                .subscribe(object : Observer<BaseBackData> {
                    override fun onSubscribe(d: Disposable) {
                        addDisposable(d)
                    }

                    override fun onNext(baseBackData: BaseBackData) {
                        view?.showMessage(baseBackData.message!!)
                    }

                    override fun onError(e: Throwable) {
                        Logger.e("Error : " + e)
                    }

                    override fun onComplete() {

                    }
                })
    }

    companion object {

        val TAG = "NewsPresenter"
    }
}
